File: python/dmp_state/DMPState.py
#!/usr/bin/env python2
# -*- coding: utf-8 -*-
"""
Created on Mon Oct 30 19:00:00 2017

@author: gsutanto
"""

import numpy as np
import os
import sys
import copy

class DMPState:
    'Base class for DMP states.'
    
    def __init__(self, X_init=None, Xd_init=None, Xdd_init=None, time_init=np.zeros((1,1)), name=""):
        self.name = name
        self.time = time_init
        if ((X_init is None) and (Xd_init is None) and (Xdd_init is None)):
            self.dmp_num_dimensions = 0
            self.X = np.empty((0,0))
            self.Xd = np.empty((0,0))
            self.Xdd = np.empty((0,0))
        else:
            if (X_init is not None):
                self.dmp_num_dimensions = X_init.shape[0]
                self.X = X_init
            if (Xd_init is not None):
                assert (self.dmp_num_dimensions == Xd_init.shape[0]), "Dimension Xd_init.shape[0]=" + str(Xd_init.shape[0]) + " is mis-matched with self.dmp_num_dimensions=" + str(self.dmp_num_dimensions) + "!"
                self.Xd = Xd_init
            else:
                self.Xd = np.zeros((self.dmp_num_dimensions,1))
            if (Xdd_init is not None):
                assert (self.dmp_num_dimensions == Xdd_init.shape[0]), "Dimension Xdd_init.shape[0]=" + str(Xdd_init.shape[0]) + " is mis-matched with self.dmp_num_dimensions=" + str(self.dmp_num_dimensions) + "!"
                self.Xdd = Xdd_init
            else:
                self.Xdd = np.zeros((self.dmp_num_dimensions,1))
            if (len(self.X.shape) == 1):
                self.X.reshape((self.dmp_num_dimensions, 1))
                self.Xd.reshape((self.dmp_num_dimensions, 1))
                self.Xdd.reshape((self.dmp_num_dimensions, 1))
                self.time.reshape((1, 1))
    
    def isValid(self):
        assert (self.X.shape[0] == self.dmp_num_dimensions), "Dimension self.X.shape[0]=" + str(self.X.shape[0]) + " is mis-matched with self.dmp_num_dimensions=" + str(self.dmp_num_dimensions) + "!"
        assert (self.Xd.shape[0] == self.dmp_num_dimensions), "Dimension self.Xd.shape[0]=" + str(self.Xd.shape[0]) + " is mis-matched with self.dmp_num_dimensions=" + str(self.dmp_num_dimensions) + "!"
        assert (self.Xdd.shape[0] == self.dmp_num_dimensions), "Dimension self.Xdd.shape[0]=" + str(self.Xdd.shape[0]) + " is mis-matched with self.dmp_num_dimensions=" + str(self.dmp_num_dimensions) + "!"
        assert (self.X.shape[1] == self.Xd.shape[1])
        assert (self.X.shape[1] == self.Xdd.shape[1])
        assert (self.X.shape[1] == self.time.shape[1])
        assert (np.amin(self.time) >= 0.0), "min(self.time)=" + str(np.amin(self.time)) + " < 0.0 (invalid!)"
        return True
    
    def getX(self):
        return copy.copy(self.X)
    
    def getXd(self):
        return copy.copy(self.Xd)
    
    def getXdd(self):
        return copy.copy(self.Xdd)
    
    def getTime(self):
        return copy.copy(self.time)
    
    def setX(self, new_X):
        assert (self.isValid()), "Pre-condition(s) checking is failed: this DMPState is invalid!"
        assert (self.dmp_num_dimensions == new_X.shape[0]), "self.dmp_num_dimensions=" + str(self.dmp_num_dimensions) + " is mis-matched with new_X.shape[0]=" + str(new_X.shape[0]) + "!"
        self.X = copy.copy(new_X)
        assert (self.isValid()), "Post-condition(s) checking is failed: this DMPState became invalid!"
        return None
    
    def setXd(self, new_Xd):
        assert (self.isValid()), "Pre-condition(s) checking is failed: this DMPState is invalid!"
        assert (self.dmp_num_dimensions == new_Xd.shape[0]), "self.dmp_num_dimensions=" + str(self.dmp_num_dimensions) + " is mis-matched with new_Xd.shape[0]=" + str(new_Xd.shape[0]) + "!"
        self.Xd = copy.copy(new_Xd)
        assert (self.isValid()), "Post-condition(s) checking is failed: this DMPState became invalid!"
        return None
    
    def setXdd(self, new_Xdd):
        assert (self.isValid()), "Pre-condition(s) checking is failed: this DMPState is invalid!"
        assert (self.dmp_num_dimensions == new_Xdd.shape[0]), "self.dmp_num_dimensions=" + str(self.dmp_num_dimensions) + " is mis-matched with new_Xdd.shape[0]=" + str(new_Xdd.shape[0]) + "!"
        self.Xdd = copy.copy(new_Xdd)
        assert (self.isValid()), "Post-condition(s) checking is failed: this DMPState became invalid!"
        return None
    
    def setTime(self, new_time):
        assert (self.isValid()), "Pre-condition(s) checking is failed: this DMPState is invalid!"
        assert (np.amin(new_time) >= 0.0), "min(new_time)=" + str(np.amin(new_time)) + " < 0.0 (invalid!)"
        self.time = new_time
        assert (self.isValid()), "Post-condition(s) checking is failed: this DMPState became invalid!"
        return None
    
    def getLength(self):
        assert (self.isValid())
        return self.time.shape[1]

File: src/dmp_state/DMPState.cpp
#include "dmp/dmp_state/DMPState.h"

namespace dmp {

DMPState::DMPState() {
  x = ZeroVectorN(MAX_DMP_NUM_DIMENSIONS);
  xd = ZeroVectorN(MAX_DMP_NUM_DIMENSIONS);
  xdd = ZeroVectorN(MAX_DMP_NUM_DIMENSIONS);
  time = 0.0;
  dmp_num_dimensions = 0;
  x.resize(dmp_num_dimensions);
  xd.resize(dmp_num_dimensions);
  xdd.resize(dmp_num_dimensions);
  rt_assertor = nullptr;
}

DMPState::DMPState(RealTimeAssertor* real_time_assertor) {
  x = ZeroVectorN(MAX_DMP_NUM_DIMENSIONS);
  xd = ZeroVectorN(MAX_DMP_NUM_DIMENSIONS);
  xdd = ZeroVectorN(MAX_DMP_NUM_DIMENSIONS);
  time = 0.0;
  dmp_num_dimensions = 0;
  x.resize(dmp_num_dimensions);
  xd.resize(dmp_num_dimensions);
  xdd.resize(dmp_num_dimensions);
  rt_assertor = real_time_assertor;
}

DMPState::DMPState(uint dmp_num_dimensions_init,
                   RealTimeAssertor* real_time_assertor) {
  dmp_num_dimensions = dmp_num_dimensions_init;
  x.resize(dmp_num_dimensions);
  xd.resize(dmp_num_dimensions);
  xdd.resize(dmp_num_dimensions);

  x = ZeroVectorN(dmp_num_dimensions);
  xd = ZeroVectorN(dmp_num_dimensions);
  xdd = ZeroVectorN(dmp_num_dimensions);
  time = 0.0;
  rt_assertor = real_time_assertor;
}

DMPState::DMPState(const VectorN& x_init,
                   RealTimeAssertor* real_time_assertor) {
  dmp_num_dimensions = x_init.rows();
  x.resize(dmp_num_dimensions);
  xd.resize(dmp_num_dimensions);
  xdd.resize(dmp_num_dimensions);

  x = x_init;
  xd = ZeroVectorN(dmp_num_dimensions);
  xdd = ZeroVectorN(dmp_num_dimensions);
  time = 0.0;
  rt_assertor = real_time_assertor;
}

DMPState::DMPState(const VectorN& x_init, const VectorN& xd_init,
                   const VectorN& xdd_init,
                   RealTimeAssertor* real_time_assertor) {
  dmp_num_dimensions = x_init.rows();
  x.resize(dmp_num_dimensions);
  xd.resize(dmp_num_dimensions);
  xdd.resize(dmp_num_dimensions);

  x = x_init;
  xd = xd_init;
  xdd = xdd_init;
  time = 0.0;
  rt_assertor = real_time_assertor;
}

DMPState::DMPState(const VectorN& x_init, const VectorN& xd_init,
                   const VectorN& xdd_init, double time_init,
                   RealTimeAssertor* real_time_assertor) {
  dmp_num_dimensions = x_init.rows();
  x.resize(dmp_num_dimensions);
  xd.resize(dmp_num_dimensions);
  xdd.resize(dmp_num_dimensions);

  x = x_init;
  xd = xd_init;
  xdd = xdd_init;
  time = time_init;
  rt_assertor = real_time_assertor;
}

DMPState::DMPState(const DMPState& other_dmpstate) {
  dmp_num_dimensions = other_dmpstate.dmp_num_dimensions;
  x.resize(dmp_num_dimensions);
  xd.resize(dmp_num_dimensions);
  xdd.resize(dmp_num_dimensions);

  x = other_dmpstate.x;
  xd = other_dmpstate.xd;
  xdd = other_dmpstate.xdd;
  time = other_dmpstate.time;
  rt_assertor = other_dmpstate.rt_assertor;
}

bool DMPState::isValid() const {
  if (rt_assert(x.rows() == dmp_num_dimensions) == false) {
    return false;
  }
  if (rt_assert(xd.rows() == dmp_num_dimensions) == false) {
    return false;
  }
  if (rt_assert(xdd.rows() == dmp_num_dimensions) == false) {
    return false;
  }
  if (rt_assert(time >= 0.0) == false) {
    return false;
  }
  return true;
}

DMPState& DMPState::operator=(const DMPState& other_dmpstate) {
  dmp_num_dimensions = other_dmpstate.dmp_num_dimensions;
  x.resize(dmp_num_dimensions);
  xd.resize(dmp_num_dimensions);
  xdd.resize(dmp_num_dimensions);

  x = other_dmpstate.x;
  xd = other_dmpstate.xd;
  xdd = other_dmpstate.xdd;
  time = other_dmpstate.time;
  rt_assertor = other_dmpstate.rt_assertor;
  return (*this);
}

uint DMPState::getDMPNumDimensions() const { return dmp_num_dimensions; }

VectorN DMPState::getX() const { return x; }

VectorN DMPState::getXd() const { return xd; }

VectorN DMPState::getXdd() const { return xdd; }

double DMPState::getTime() const { return time; }

bool DMPState::setX(const VectorN& new_x) {
  // pre-condition(s) checking
  if (rt_assert(DMPState::isValid()) == false) {
    return false;
  }

  if (rt_assert(dmp_num_dimensions == new_x.rows())) {
    x = new_x;
    return true;
  } else {
    return false;
  }
}

bool DMPState::setXd(const VectorN& new_xd) {
  // pre-condition(s) checking
  if (rt_assert(DMPState::isValid()) == false) {
    return false;
  }

  if (rt_assert(dmp_num_dimensions == new_xd.rows())) {
    xd = new_xd;
    return true;
  } else {
    return false;
  }
}

bool DMPState::setXdd(const VectorN& new_xdd) {
  // pre-condition(s) checking
  if (rt_assert(DMPState::isValid()) == false) {
    return false;
  }

  if (rt_assert(dmp_num_dimensions == new_xdd.rows())) {
    xdd = new_xdd;
    return true;
  } else {
    return false;
  }
}

bool DMPState::setTime(double new_time) {
  // pre-condition(s) checking
  if (rt_assert(DMPState::isValid()) == false) {
    return false;
  }

  if (rt_assert(new_time >= 0.0)) {
    time = new_time;
    return true;
  } else {
    return false;
  }
}

DMPState::~DMPState() {}

}  // namespace dmp
